File: Scada-Systems/openAPC/ControlRoom/hmiObject.cpp
/******************************************************************************

This file is part of ControlRoom process control/HMI software.

ControlRoom is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.

ControlRoom is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License along with
ControlRoom. If not, see http://www.gnu.org/licenses/

*******************************************************************************

For different licensing and/or usage of the sources apart from GPL or any other
open source license, please contact us at https://openapc.com/contact.php

*******************************************************************************/

#ifdef ENV_INT
 #include "../hmiObject.cpp"
#else //ENV_INT

#include <wx/wx.h>
#include <wx/font.h>
#include <wx/thread.h>
#include <wx/dcbuffer.h>

#ifndef __WXMSW__
#include <arpa/inet.h>
#endif

#include "iff.h"
#include "globals.h"
#include "MainApp.h"
#include "flowObject.h"
#include "hmiObject.h"
#include "hmiSimpleButton.h"
#include "HMICanvas.h"
#include "hmiFloatField.h"
#include "hmiHGauge.h"
#include "hmiHSlider.h"
#include "hmiImage.h"
#include "hmiPrimitive.h"
#include "hmiImageButton.h"
#include "hmiToggleImageButton.h"
#include "hmiLine.h"
#include "hmiNumField.h"
#include "hmiSimpleButton.h"
#include "hmiTextLabel.h"
#include "hmiToggleButton.h"
#include "hmiRadioButton.h"
#include "hmiTextField.h"
#include "hmiVGauge.h"
#include "hmiVSlider.h"
#include "hmiTabbedPane.h"
#include "hmiStackedPane.h"
#include "hmiAdditionalPane.h"
#include "hmiSinglePanel.h"
#include "hmiAngularMeter.h"
#include "hmiAngularRegulator.h"
#include "hmiLCDNumber.h"
#include "hmiSpecialPanel.h"
#include "hmiExternalIOLib.h"


hmiObject::hmiObject(BasePanel *parent,flowObject *obj)
          :flowObject(obj)
          , uiElement(NULL)
          , returnOK(false)
#ifndef ENV_EDITOR
          , logDataCtrl(NULL)
#endif
          , lastFont_pointSize(-1)
          , lastFont_style(-1)
          , lastFont_weight(-1)
          , lastFont_faceName(_T(""))
          , m_loadError(OAPC_OK)
{
   m_parent=parent;
#ifdef ENV_EDITOR
   m_userPriviEnable=OAPC_USERPRIVI_DISABLE|OAPC_USERPRIVI_HIDE;
   noUIFlags=0xFFFFFFFF;
   noFlowFlags=0;
   data.reservedLayoutAlignment=0;
   data.hmiFlags=0;
#endif
   if (m_parent) data.isChildOf=m_parent->m_myID; // no parent in case of silent load
   returnOK=true;
}


hmiObject::~hmiObject()
{
   hmiMutex *mutex;

   mutex=g_objectList.getMutexByObjectID(this->data.id);
   if (mutex) mutex->removeObjectID(this->data.id);
#ifdef ENV_PLAYER
   if (logDataCtrl) delete logDataCtrl;
#endif
   if ((uiElement) && (m_doDestroy))
   {
#ifdef ENV_PLAYER
      uiElement->Close();
      uiElement->Destroy();
#else
      uiElement->Show(false); // just hide it, deletion out of an event handler would cause a crash
#endif
   }
}


void hmiObject::OnValueChanged(wxCommandEvent &event)
{
   if (event.GetId()==data.type)
    doApplyData(event.GetEventObject()!=NULL);

   if (event.GetEventObject()) delete event.GetEventObject();
}


void hmiObject::triggerGUIAction(const int id,const wxByte all)
{
   wxCommandEvent* evt = new wxCommandEvent();

   if (all) evt->SetEventObject(new wxObject());
   else evt->SetEventObject(NULL);
   evt->SetEventType(wxEVT_COMMAND_BUTTON_CLICKED);
   evt->SetId(id);
   AddPendingEvent(*evt);
}


void hmiObject::setParent(BasePanel *parent)
{
   m_parent=parent;
   data.isChildOf=parent->m_myID;
}


void hmiObject::doApplyData(const wxByte WXUNUSED(all))
{
}


void hmiObject::applyData(wxByte all)
{
   triggerGUIAction(data.type,all);
}


bool hmiObject::applyLayoutData()
{
   return false;
}

void hmiObject::drawObject(wxAutoBufferedPaintDC* /*dc*/)
{
}


wxWindow *hmiObject::createUIElement()
{
   if (!uiElement) return NULL;
   applyData(1);

#ifndef ENV_PLAYER
   uiElement->Connect(wxEVT_RIGHT_UP,wxMouseEventHandler(HMICanvas::OnChildMouseEvent),NULL,m_parent);
   uiElement->Connect(wxEVT_MOTION,wxMouseEventHandler(HMICanvas::OnChildMouseEvent),NULL,m_parent);
   uiElement->Connect(wxEVT_LEFT_DOWN,wxMouseEventHandler(HMICanvas::OnChildMouseEvent),NULL,m_parent);
   uiElement->Connect(wxEVT_LEFT_UP,wxMouseEventHandler(HMICanvas::OnChildMouseEvent),NULL,m_parent);
   uiElement->Connect(wxEVT_LEFT_DCLICK,wxMouseEventHandler(HMICanvas::OnChildMouseEvent),NULL,m_parent);
#endif

   return uiElement;
}



void hmiObject::applyFont(wxWindow *ui)
{
   if ((lastFont_pointSize!=data.font[digi[0]].pointSize) ||
       (lastFont_style    !=data.font[digi[0]].style) ||
       (lastFont_weight   !=data.font[digi[0]].weight) ||
       (lastFont_faceName.Cmp(fontFaceName[digi[0]])!=0))
   {
#if wxCHECK_VERSION(2,9,0)
      g_setFont(ui,wxFont(data.font[digi[0]].pointSize,
                          wxFONTFAMILY_DEFAULT,
                          (wxFontStyle)data.font[digi[0]].style,
                          (wxFontWeight)data.font[digi[0]].weight,
                          false,
                          fontFaceName[digi[0]]));
#else
      g_setFont(ui,wxFont(data.font[digi[0]].pointSize,
                          wxFONTFAMILY_DEFAULT,
                          data.font[digi[0]].style,
                          data.font[digi[0]].weight,
                          false,
                          fontFaceName[digi[0]]));
#endif

      lastFont_pointSize=data.font[digi[0]].pointSize;
      lastFont_style=data.font[digi[0]].style;
      lastFont_weight=data.font[digi[0]].weight;
      lastFont_faceName=fontFaceName[digi[0]];
   }
}


wxInt32 hmiObject::getLoadError()
{
   return m_loadError;
}


wxInt32 hmiObject::load(wxFile *FHandle,wxUint32 chunkSize,wxUint32 IDOffset)
{
   wxInt32 res;

   res=flowObject::load(FHandle,chunkSize,IDOffset);
#ifdef ENV_PLAYER
   logDataCtrl=new LogDataCtrl(data.log);
#endif

   if ((data.hmiFlags & 0xFFFFFFFE)!=0) data.hmiFlags=0; // for compatibility with old projects where hmiFlags does not has been initialised correctly
   return res;
}



#ifdef ENV_EDITOR
hmiObject *hmiObject::duplicate()
{
    hmiObject *dupObject=NULL;
    if (data.type==HMI_TYPE_FLOATFIELD)             dupObject=new hmiFloatField(m_parent,this);
    else if (data.type==HMI_TYPE_HGAUGE)            dupObject=new hmiHGauge(m_parent,this);
    else if (data.type==HMI_TYPE_HSLIDER)           dupObject=new hmiHSlider(m_parent,this);
    else if (data.type==HMI_TYPE_IMAGE)             dupObject=new hmiImage(m_parent,this);
    else if ((data.type==HMI_TYPE_FREELINE) || (data.type==HMI_TYPE_FRAME) || (data.type==HMI_TYPE_ELLIPSE))
                                                    dupObject=new hmiPrimitive(m_parent,this,data.type);
    else if (data.type==HMI_TYPE_IMAGEBUTTON)       dupObject=new hmiImageButton(m_parent,this);
    else if (data.type==HMI_TYPE_TOGGLEIMAGEBUTTON) dupObject=new hmiToggleImageButton(m_parent,this);
    else if (data.type==HMI_TYPE_LINE)              dupObject=new hmiLine(m_parent,this);
    else if (data.type==HMI_TYPE_NUMFIELD)          dupObject=new hmiNumField(m_parent,this);
    else if (data.type==HMI_TYPE_SIMPLEBUTTON)      dupObject=new hmiSimpleButton(m_parent,this);
    else if (data.type==HMI_TYPE_TEXTLABEL)         dupObject=new hmiTextLabel(m_parent,this);
    else if (data.type==HMI_TYPE_TOGGLEBUTTON)      dupObject=new hmiToggleButton(m_parent,this);
    else if (data.type==HMI_TYPE_RADIOBUTTON)       dupObject=new hmiRadioButton(m_parent,this,HMI_TYPE_RADIOBUTTON);
    else if (data.type==HMI_TYPE_CHECKBOX)          dupObject=new hmiRadioButton(m_parent,this,HMI_TYPE_CHECKBOX);
    else if (data.type==HMI_TYPE_TEXTFIELD)         dupObject=new hmiTextField(m_parent,this,HMI_TYPE_TEXTFIELD);
    else if (data.type==HMI_TYPE_PASSWORDFIELD)     dupObject=new hmiTextField(m_parent,this,HMI_TYPE_PASSWORDFIELD);
    else if (data.type==HMI_TYPE_VGAUGE)            dupObject=new hmiVGauge(m_parent,this);
    else if (data.type==HMI_TYPE_VSLIDER)           dupObject=new hmiVSlider(m_parent,this);
    else if (data.type==HMI_TYPE_TABBEDPANE)        dupObject=new hmiTabbedPane(m_parent,this);
    else if (data.type==HMI_TYPE_STACKEDPANE)       dupObject=new hmiStackedPane(m_parent,this);
    else if (data.type==HMI_TYPE_ADDITIONALPANE)    dupObject=new hmiAdditionalPane(m_parent,this);
    else if (data.type==HMI_TYPE_SINGLEPANEL)       dupObject=new hmiSinglePanel(m_parent,this);
    else if (data.type==HMI_TYPE_ANGULARMETER)      dupObject=new hmiAngularMeter(m_parent,this);
    else if (data.type==HMI_TYPE_LCDNUMBER)         dupObject=new hmiLCDNumber(m_parent,this);
    else if (data.type==HMI_TYPE_ANGULARREGULATOR)  dupObject=new hmiAngularRegulator(m_parent,this);
    else if (data.type==HMI_TYPE_SP_USERMANAGEMENT) dupObject=new hmiSpecialPanel(m_parent,this,HMI_TYPE_SP_USERMANAGEMENT);
    else if (data.type==HMI_TYPE_EXTERNAL_LIB)
    {
       dupObject=new hmiExternalIOLib(((hmiExternalIOLib*)this)->lib,m_parent,this,0);
    }
    wxASSERT(dupObject);
    if (!dupObject) return NULL;
    dupObject->createUIElement();
    dupObject->setData(this);

    return dupObject;
}



wxString hmiObject::getDefaultName()
{
    wxASSERT(0);
    return _T("##invalid hmiObject##");
}



wxPoint hmiObject::getMinSize()
{
   return wxPoint(10,10);
}


wxPoint hmiObject::getMaxSize()
{
   return wxPoint(10000,10000);
}


wxPoint hmiObject::getDefSize()
{
    return wxPoint(40,20);
}
#endif



void hmiObject::setSize(wxRealPoint msize)
{
   data.msizex=msize.x;
   data.msizey=msize.y;
   if ((uiElement) && (data.msizex>0))
    uiElement->SetSize(msize.x/1000.0,msize.y/1000.0);
}



wxRealPoint hmiObject::getSize()
{
   return wxRealPoint(data.msizex,data.msizey);
}



void hmiObject::setPos(wxRealPoint pos)
{
   data.mposx=pos.x;
   data.mposy=pos.y;
   if ((uiElement) && (data.msizex>0))
    uiElement->SetPosition(wxPoint(pos.x/1000.0,pos.y/1000.0));
}



wxRealPoint hmiObject::getPos()
{
   return wxRealPoint(data.mposx,data.mposy);
}



#ifdef ENV_PLAYER
void hmiObject::setUIElementState(wxInt32 state)
{
   wxASSERT(uiElement);
   if (!uiElement) return;
   if (state==2)
   {
      if (!uiElement->IsShown()) uiElement->Show(true);
      if (!uiElement->IsEnabled()) uiElement->Enable(true);
   }
   else if (state==1)
   {
      if (!uiElement->IsShown()) uiElement->Show(true);
      if (uiElement->IsEnabled()) uiElement->Enable(false);
   }
   else if (state==0) uiElement->Show(false);
}
#endif



#ifdef ENV_EDITOR
wxPanel* hmiObject::getConfigPanel(wxString *WXUNUSED(name),wxWindow *WXUNUSED(parent),wxInt32 WXUNUSED(num))
{
    return NULL;
}



void hmiObject::setConfigData()
{
}


wxInt32 hmiObject::saveSOBJ(wxFile *FHandle,bool isCompiled)
{
   struct hmiFlowData             flowData;
   wxList::compatibility_iterator node;
   FlowConnection                *connection;
   wxInt32                        length;

   if (!FHandle) return 0;
   if (flowList.GetCount()<0) return 0;
   checkConnections();
   FHandle->Write(CHUNK_FLOW_SOBJ,4);
   if (isCompiled) length=sizeof(struct hmiFlowData)+(flowList.GetCount()*FLOW_CONNECTION_HEAD_SIZE);
   else length=sizeof(struct hmiFlowData)+(flowList.GetCount()*sizeof(struct flowConnectionData));
   length=htonl(length);
   FHandle->Write(&length,4);

   wxASSERT(data.id>0);
   flowData.id=htonl(data.id);
   flowData.version=htonl(1);
   flowData.usedFlows=htonl(flowList.GetCount());
   flowData.maxEdges=htonl(MAX_CONNECTION_EDGES);
   flowData.flowFlags=htonl(m_flowFlags);
   length=FHandle->Write(&flowData,sizeof(struct hmiFlowData));

   node=flowList.GetFirst();
   while (node)
   {
      connection=(FlowConnection*)node->GetData();
      length+=connection->saveFlow(FHandle,isCompiled);
      node=node->GetNext();
   }
   return length;
}
#else



wxUint64 hmiObject::getAssignedOutput(wxUint64 input)
{
   if (input==OAPC_DIGI_IO0) return OAPC_DIGI_IO0;
   return 0;
}



void hmiObject::setSelected(bool WXUNUSED(selected))
{
}



void hmiObject::updateSelectionState()
{
}
#endif

#endif //ENV_INT

File: Scada-Systems/openAPC/AndroidViewer/src/com/oapc/hmiObject.java
package com.oapc;

import android.view.*;
import android.widget.*;

import com.oapc.iserver.*;
import com.oapc.liboapc.*;
import android.os.*;



public class hmiObject extends flowObject 
{
   public static final int HMI_TYPE_MASK              =0x000000FF;
   public static final int HMI_TYPE_SIMPLEBUTTON      =0x00000001;
   public static final int HMI_TYPE_TOGGLEBUTTON      =0x00000002;
   public static final int HMI_TYPE_HSLIDER           =0x00000003;
   public static final int HMI_TYPE_VSLIDER           =0x00000004;
   public static final int HMI_TYPE_TEXTLABEL         =0x00000005;
   public static final int HMI_TYPE_LINE              =0x00000006;
   public static final int HMI_TYPE_IMAGEBUTTON       =0x00000007;
   public static final int HMI_TYPE_IMAGE             =0x00000008;
   public static final int HMI_TYPE_HGAUGE            =0x00000009;
   public static final int HMI_TYPE_VGAUGE            =0x0000000A;
   public static final int HMI_TYPE_NUMFIELD          =0x0000000B;
   public static final int HMI_TYPE_FLOATFIELD        =0x0000000C;
   public static final int HMI_TYPE_TEXTFIELD         =0x0000000D;
   public static final int HMI_TYPE_TABBEDPANE        =0x0000000E;
   public static final int HMI_TYPE_STACKEDPANE       =0x0000000F;
   public static final int HMI_TYPE_ADDITIONALPANE    =0x00000010;
   public static final int HMI_TYPE_TOGGLEIMAGEBUTTON =0x00000011;
   public static final int HMI_TYPE_ANGULARMETER      =0x00000012;
   public static final int HMI_TYPE_LCDNUMBER         =0x00000013;
   public static final int HMI_TYPE_ANGULARREGULATOR  =0x00000014;
   public static final int HMI_TYPE_EXTERNAL_LIB      =0x00000015;
   public static final int HMI_TYPE_SINGLEPANEL       =0x00000016;
   public static final int HMI_TYPE_RADIOBUTTON       =0x00000017;
   public static final int HMI_TYPE_CHECKBOX          =0x00000018;
   public static final int HMI_TYPE_SP_USERMANAGEMENT =0x00000019;
   public static final int HMI_TYPE_PASSWORDFIELD     =0x0000001A;

   public static final int HMI_STATE_FLAG_DISABLED =0x0001;
   public static final int HMI_STATE_FLAG_RO       =0x0004;

   public static final int HMI_FLOWFLAG_ON_DIGI_IN1_HI_OUT6 =0x0001; /** trigger data output on OUT 6 in case of a signal on digital input 0*/
   public static final int HMI_FLOWFLAG_ON_DIGI_IN1_LO_OUT6 =0x0002; /** trigger data output on OUT 6 in case of a signal on digital input 0*/
   public static final int HMI_FLOWFLAG_ON_DIGI_IN1_HI_OUT7 =0x0004; /** trigger data output on OUT 7 in case of a signal on digital input 0*/
   public static final int HMI_FLOWFLAG_ON_DIGI_IN1_LO_OUT7 =0x0008; /** trigger data output on OUT 7 in case of a signal on digital input 0*/
   public static final int HMI_FLOWFLAG_ON_CHANGING_OUT6    =0x0010; /** trigger data output on OUT 6 in case of value is changing actually (includes HMI_FLOWFLAG_ON_CHANGED); in case of sliders or permanently changing data sources this may cause massive data transmission*/
   public static final int HMI_FLOWFLAG_ON_CHANGED_OUT6     =0x0020; /** trigger data output on OUT 6 in case of value was changed */
   public static final int HMI_FLOWFLAG_ON_CHANGING_OUT7    =0x0040; /** trigger data output on OUT 7 in case of value is changing actually (includes HMI_FLOWFLAG_ON_CHANGED); in case of sliders or permanently changing data sources this may cause massive data transmission*/
   public static final int HMI_FLOWFLAG_ON_CHANGED_OUT7     =0x0080; /** trigger data output on OUT 7 in case of value was changed */

   public static final int HMI_NOFLOWFLAG_ON_DIGI_IN1 =0x0001; /** do not give option to react on digital input 1 */
   public static final int HMI_NOFLOWFLAG_ON_CHANGING =0x0002; /** do not give option to react on changing value */
   public static final int HMI_NOFLOWFLAG_ON_CHANGED  =0x0004; /** do not give option to react on changed value */
   public static final int HMI_NOFLOWFLAG_OUT6        =0x0008; /** do not give options for output 6*/

   View            uiElement;
   BasePanel       m_parent;
   boolean         returnOK;
   int             m_mutexID;
   
   LogDataCtrl     logDataCtrl;

   int             lastFont_pointSize,lastFont_style,lastFont_weight;
   String          lastFont_faceName;   
   
   hmiObject(BasePanel parent,flowObject obj)
   {
      super(obj);
	  returnOK=false;
	  m_parent=parent;
	  if (m_parent!=null) data.isChildOf=m_parent.m_myID; // no parent in case of silent load
	  uiElement=null;
	  lastFont_pointSize=-1;
      lastFont_style=-1;
	  lastFont_weight=-1;
	  lastFont_faceName="";
	  logDataCtrl=new LogDataCtrl(data.log);
	  returnOK=true;
	}
	
	
	
   void close()
   {
/* TODO:      hmiMutex *mutex;
	
	  mutex=g_objectList.getMutexByObjectID(this->data.id);
	  if (mutex) mutex->removeObjectID(this->data.id);*/
	  if (logDataCtrl!=null) logDataCtrl.close();
	  logDataCtrl=null;
/* TODO:     if ((uiElement) && (m_doDestroy))
	  {
         uiElement->Close();
         uiElement->Destroy();
      }*/
   }
	
	
	
   void setParent(BasePanel parent)
   {
      m_parent=parent;
      data.isChildOf=parent.m_myID;
   }

   
   
   void applyData(boolean all)
   {
/** custom background colour not supported in first versions -> damages button layout, no smooth possibility to set background
 *       uiElement.setBackgroundColor(wxConvert.fromWxColour(data.bgcolor[digi[0]]));
 */
	  super.applyData(all);
   }



   View createUIElement()
   {
      if (uiElement==null) return null;
      m_parent.absLayout.addView(uiElement,new AbsoluteLayout.LayoutParams(data.msizex/1000,data.msizey/1000,data.mposx/1000,data.mposy/1000));      
      return uiElement;
   }   
   

   
   void applyFont(View ui)
   {
/* custom fonts not supported in first versions
       lastFont_pointSize=data.font[digi[0]].pointSize;
      lastFont_style=data.font[digi[0]].style;
      lastFont_weight=data.font[digi[0]].weight;
      lastFont_faceName=data.font[digi[0]].fontFaceName;*/
   }
   
   

   long getAssignedOutput(long input)
   {
      if (input==OAPCIServer.OAPC_DIGI_IO0) return OAPCIServer.OAPC_DIGI_IO0;
      return 0;
   }

   
   
   void updateSelectionState()
   {
   }   

   
   
   void setSelected(boolean selected)
   {
   }
   
   
}
